### ISSUE TITLE
filetype.nvim breaks ftdetect

### ISSUE THREAD

**Issue by gegoune** [nathom/filetype.nvim#9]
Nice plugin, thanks!

Given `/usr/local/Cellar/neovim/<ver>/share/nvim/runtime/filetype.vim` is not fully loaded because of `vim.g.did_load_filetypes = 1` it also does not load 

```viml
    " Use the filetype detect plugins.  They may overrule any of the previously
    " detected filetypes.
    runtime! ftdetect/*.vim
    runtime! ftdetect/*.lua
```
which is part of it. This breaks user's custom filetypes specs from `ftdetect/` directory.

Do you think that

1. your plugin could run those two commands
2. provide configuration option (setup function?) to add custom mappings to extensions

Thanks!

**Comment by nathom** [comment]
I think adding support for user defined filetypes is a good idea. However, I won't include sourcing other `ftdetect` files in the plugin since its purpose is to avoid creating autocommands.

**Comment by gegoune** [comment]
Also there should be a way to override default mappings. What to you think?

**Comment by timbedard** [comment]
Oh dang, yeah. Just realized this. I have a few custom ftdetect autocommands and I would love to replace them with additional configuration for this plugin. I moved them to a file that I know will be sourced for now.

**Comment by gegoune** [comment]
@nathom Just checking out your dev branch. Looking good, thanks lot!

One problem I am seeing is my custom literal override:

```lua
require('filetype').setup {
  overrides = {
    literal = {
      ['something.log'] = 'somethinglog',
    },
    complex = {
      ['.*zsh/fpath/'] = 'zsh',
      ['.*ssh/config'] = 'sshconfig',
    },
  },
}
```

When I open `something.log` file it gets `log` filetype set.

**Comment by Furkanzmc** [comment]
This issue is one of the things that was stopping me from using the otherwise great plugin. I have one question regarding the fix, how does it help a plugin developer? If I install a plugin, and it has its own file type detection, do I have to manually enter the filetypes in to the new setup function call?

**Comment by ruifm** [comment]
I personally just have this at the top of my `init.lua`:

```lua
vim.g.did_load_filetypes = 1
vim.cmd([[ runtime! ftdetect/*.vim]])
vim.cmd([[runtime! ftdetect/*.lua]])
```

I know the point of the plugin is to avoid loading more than 1 filetype setting autocmd, but the vast majority is still in `filetype.vim`. With this plugin, and the lines above, I still get the startup performance beneficts + external `ftdetect` plugins (and my own `ftdetect`) still work as expected.

I can't imagine legacy filetype plugin authors ever catering to `filetype.nvim` honestly.

**Comment by nathom** [comment]
It seems that this is an ordering issue. The regexes always run after the extension and literal tables, so they will get overridden. Maybe I'll need to keep a separate overrides table that has precedence over the rest.

**Comment by gegoune** [comment]
> One problem I am seeing is my custom literal override:
> 
> ```lua
> require('filetype').setup {
>   overrides = {
>     literal = {
>       ['something.log'] = 'somethinglog',
>     },
>     complex = {
>       ['.*zsh/fpath/'] = 'zsh',
>       ['.*ssh/config'] = 'sshconfig',
>     },
>   },
> }
> ```
> 
> When I open `something.log` file it gets `log` filetype set.

It seems to be fixed in main now. And since there is a way to provide overrides (no need for ftdetect, unless some other plugins/etc need it) I will close that issue as it's all working for me now.

Can you see any issues with not having those `runtime!` on ftdetect files?

**Comment by gegoune** [comment]
Reopening as per https://github.com/nathom/filetype.nvim/issues/41#issuecomment-963032730

**Comment by jemag** [comment]
I strongly agree with @clason , this has broken the many filetype plugins that I use, and I think the general assumption would be that these would still work. I think a mention of this issue in the README in the meantime would also be useful.

**Comment by ajukraine** [comment]
Unfortunately it also break https://github.com/kassio/neoterm, because it relies on 'ftdetect' folder for its core logic.

**Comment by clason** [comment]
**PSA:** `filetype.lua` is now the default on Neovim master (and 0.8.0, when that is released). It has full parity with `filetype.vim` (including `scripts.vim` and `setf FALLBACK` support), passing Vim's filetype detection test suite. It will be kept in sync with future filetype detection changes in Vim and should thus be a safe and non-breaking replacement.

That means this plugin is free to take a more opinionated approach without concern for backward compatibility, if desired. (Although it would be good to keep `filetype.lua` enabled as a fallback, as this doesn't have the same performance penalty as `filetype.vim`.)
